exp/runtime/api/v1beta1/extensionconfig_types.go

```diff
@@ -0,0 +1,195 @@
+	//
+	// The scheme should be "https"; the URL should begin with "https://".
+	// "http" is supported for insecure development purposes only.
+	//
+	// A path is optional, and if present may be any string permissible in
+	// a URL. If a path is set it will be used as prefix and the hook-specific
+	// path will be appended.
+	//
+	// Attempting to use a user or basic auth e.g. "user:password@" is not
+	// allowed. Fragments ("#...") and query parameters ("?...") are not
+	// allowed either.
+	//
+	// +optional
+	URL *string `json:"url,omitempty"`
+
+	// Service is a reference to the Kubernetes service for the ExtensionHandler.
+	// Either `service` or `url` must be specified.
+	//
+	// If the ExtensionHandler is running within a cluster, then you should use `service`.
+	//
+	// +optional
+	Service *ServiceReference `json:"service,omitempty"`
+
+	// CABundle is a PEM encoded CA bundle which will be used to validate the ExtensionHandler's server certificate.
+	// +optional
+	CABundle []byte `json:"caBundle,omitempty"`
+}
+
+// ServiceReference holds a reference to a Kubernetes Service.
+type ServiceReference struct {
+	// Namespace is the namespace of the service.
+	Namespace string `json:"namespace"`
+
+	// Name is the name of the service.
+	Name string `json:"name"`
+
+	// Path is an optional URL path which will be sent in any request to
+	// this service. If a path is set it will be used as prefix and the hook-specific
+	// path will be appended.
+	// +optional
+	Path *string `json:"path,omitempty"`
+
+	// Port is the port on the service that hosting ExtensionHandler.
+	// Default to 8443.
+	// `port` should be a valid port number (1-65535, inclusive).
+	// +optional
+	Port *int32 `json:"port,omitempty"`
+}
+
+// ANCHOR_END: ExtensionConfigSpec
+
+// ANCHOR: ExtensionConfigStatus
+
+// ExtensionConfigStatus defines the observed state of ExtensionConfig.
+type ExtensionConfigStatus struct {
+	// ExtensionHandlers defines the current ExtensionHandlers supported by an Extension.
+	// +optional
+	ExtensionHandlers []ExtensionHandler `json:"extensionHandlers,omitempty"`
+
+	// Conditions define the current service state of the ExtensionConfig.
+	// +optional
+	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+}
+
+// ExtensionHandler specifies the details of a handler for a particular runtime hook registered by an Extension server.
+type ExtensionHandler struct {
+	// Name is the unique name of the ExtensionHandler.
+	Name string `json:"name"`
+
+	// Hook defines the versioned runtime hook which this ExtensionHandler serves.
+	Hook GroupVersionHook `json:"hook"`
+
+	// TimeoutSeconds defines the timeout duration for client calls to the ExtensionHandler.
+	// +optional
+	TimeoutSeconds *int32 `json:"timeoutSeconds,omitempty"`
+
+	// FailurePolicy defines how failures in calls to the ExtensionHandler should be handled by a client.
+	// +optional
+	FailurePolicy *FailurePolicy `json:"failurePolicy,omitempty"`
```

Comment:
Do either of these need to be pointers, is there a semantic difference between the zero go value and the nil value?
If there is, it's worth adding a comment to point this out, eg, when timeoutSeconds is set to 0, this disables the timeout <- I have no idea if that's true, just an example

Reply:
I think these are pointers in line with the guidelines optional fields being pointers. Both of these will be defaulted if unset by users.

Reply:
I'm not a strong opinion on this, but, upstream kube conventions say yes, these should be pointers, the CAPI conventions are loosely worded, `Fields SHOULD be pointers if there is a good reason for it, for example:`, which to me reads as we would prefer them not to be pointers unless we have a good reason, based on that I would say these shouldn't be pointers but willing to yield as it's pretty minor